File: services/ingestion/metrics.go
package main

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	metricMessagesReceived = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "messages_received_total",
		Help: "MQTT messages received by the ingestion service, by topic kind.",
	}, []string{"topic_kind"})

	metricMessagesProcessed = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "messages_processed_total",
		Help: "Messages successfully published downstream, by result.",
	}, []string{"result"})

	metricMessagesFailed = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "messages_failed_total",
		Help: "Messages that failed processing, by reason.",
	}, []string{"reason"})

	metricDLQMessages = promauto.NewCounter(prometheus.CounterOpts{
		Name: "dlq_messages_total",
		Help: "Messages routed to the dead-letter topic.",
	})

	metricProcessingLatency = promauto.NewHistogram(prometheus.HistogramOpts{
		Name:    "processing_latency_seconds",
		Help:    "End-to-end time from MQTT receipt to Kafka publish (or dead-letter) completion.",
		Buckets: prometheus.DefBuckets,
	})

	metricMQTTConnected = promauto.NewGauge(prometheus.GaugeOpts{
		Name: "mqtt_connections",
		Help: "Whether the ingestion service currently holds a live MQTT connection (0 or 1).",
	})
)

File: services/ingestion/metrics.py
"""Prometheus metrics, the Python port of metrics.go. Same metric names and
labels, so existing Grafana dashboards/alerts keep working unchanged."""

from prometheus_client import Counter, Gauge, Histogram

messages_received_total = Counter(
    "messages_received_total",
    "MQTT messages received by the ingestion service, by topic kind.",
    ["topic_kind"],
)

messages_processed_total = Counter(
    "messages_processed_total",
    "Messages successfully published downstream, by result.",
    ["result"],
)

messages_failed_total = Counter(
    "messages_failed_total",
    "Messages that failed processing, by reason.",
    ["reason"],
)

dlq_messages_total = Counter(
    "dlq_messages_total",
    "Messages routed to the dead-letter topic.",
)

processing_latency_seconds = Histogram(
    "processing_latency_seconds",
    "End-to-end time from MQTT receipt to Kafka publish (or dead-letter) completion.",
)

mqtt_connections = Gauge(
    "mqtt_connections",
    "Whether the ingestion service currently holds a live MQTT connection (0 or 1).",
)
